File: noframework/golang/tech-writer-agent/tools.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"os"
	"path/filepath"
	"strings"
	
	gitignore "github.com/denormal/go-gitignore"
)

// Tool represents a callable tool function
type Tool struct {
	Name        string
	Description string
	Function    func(args map[string]interface{}) (interface{}, error)
}

// ToolResult represents the result of a tool call
type ToolResult struct {
	Success bool        `json:"success"`
	Result  interface{} `json:"result,omitempty"`
	Error   string      `json:"error,omitempty"`
}

// FileSearchResult represents the result of finding files
type FileSearchResult struct {
	Files []string `json:"files"`
	Count int      `json:"count"`
}

// FileReadResult represents the result of reading a file
type FileReadResult struct {
	File    string `json:"file"`
	Content string `json:"content"`
}

// Available tools
var Tools = map[string]Tool{
	"find_all_matching_files": {
		Name:        "find_all_matching_files",
		Description: "Find files matching a pattern while respecting .gitignore",
		Function:    findAllMatchingFiles,
	},
	"read_file": {
		Name:        "read_file",
		Description: "Read the contents of a file",
		Function:    readFile,
	},
}

// findAllMatchingFiles finds files matching a pattern
func findAllMatchingFiles(args map[string]interface{}) (interface{}, error) {
	// Extract arguments with defaults
	directory, ok := args["directory"].(string)
	if !ok {
		return nil, fmt.Errorf("directory parameter is required")
	}
	
	pattern, ok := args["pattern"].(string)
	if !ok {
		pattern = "*"
	}
	
	respectGitignore := true
	if val, ok := args["respect_gitignore"].(bool); ok {
		respectGitignore = val
	}
	
	includeHidden := false
	if val, ok := args["include_hidden"].(bool); ok {
		includeHidden = val
	}
	
	includeSubdirs := true
	if val, ok := args["include_subdirs"].(bool); ok {
		includeSubdirs = val
	}
	
	log.Printf("Tool invoked: find_all_matching_files(directory='%s', pattern='%s', respect_gitignore=%v, include_hidden=%v, include_subdirs=%v)",
		directory, pattern, respectGitignore, includeHidden, includeSubdirs)
	
	// Resolve directory path
	absDir, err := filepath.Abs(directory)
	if err != nil {
		return nil, fmt.Errorf("error resolving directory path: %w", err)
	}
	
	// Check if directory exists
	if _, err := os.Stat(absDir); os.IsNotExist(err) {
		log.Printf("Directory not found: %s", directory)
		return FileSearchResult{Files: []string{}, Count: 0}, nil
	}
	
	// Get gitignore matcher if needed
	var matcher gitignore.GitIgnore
	if respectGitignore {
		matcher = loadGitignoreMatcher(absDir)
	}
	
	var matchingFiles []string
	
	// Walk the directory tree
	err = filepath.Walk(absDir, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			return nil // Skip files we can't access
		}
		
		// Skip directories
		if info.IsDir() {
			// Always skip .git directory
			if filepath.Base(path) == ".git" {
				return filepath.SkipDir
			}
			// Skip subdirectories if not included
			if !includeSubdirs && path != absDir {
				return filepath.SkipDir
			}
			return nil
		}
		
		// Get relative path for pattern matching
		relPath, err := filepath.Rel(absDir, path)
		if err != nil {
			return nil
		}
		
		// Skip hidden files if not included
		if !includeHidden && strings.HasPrefix(filepath.Base(path), ".") {
			// Check if any parent directory is hidden
			parts := strings.Split(relPath, string(filepath.Separator))
			hasHiddenParent := false
			for i := 0; i < len(parts)-1; i++ { // Exclude the filename itself
				if strings.HasPrefix(parts[i], ".") {
					hasHiddenParent = true
					break
				}
			}
			// Only skip if it's in a hidden directory
			if hasHiddenParent {
				return nil
			}
			// Hidden files in non-hidden directories (like .gitignore) should be included
		}
		
		// Skip gitignored files
		if respectGitignore && shouldIgnore(relPath, matcher) {
			return nil
		}
		
		// Check if file matches pattern
		matched, err := filepath.Match(pattern, filepath.Base(path))
		if err != nil {
			return nil
		}
		
		if matched {
			matchingFiles = append(matchingFiles, path)
		}
		
		return nil
	})
	
	if err != nil {
		return nil, fmt.Errorf("error walking directory: %w", err)
	}
	
	log.Printf("Found %d matching files", len(matchingFiles))
	
	return FileSearchResult{
		Files: matchingFiles,
		Count: len(matchingFiles),
	}, nil
}

// readFile reads the contents of a file
func readFile(args map[string]interface{}) (interface{}, error) {
	filePath, ok := args["file_path"].(string)
	if !ok {
		return nil, fmt.Errorf("file_path parameter is required")
	}
	
	log.Printf("Tool invoked: read_file(file_path='%s')", filePath)
	
	// Check if file exists
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		return map[string]string{"error": fmt.Sprintf("File not found: %s", filePath)}, nil
	}
	
	// Check if it's a binary file
	if isBinary(filePath) {
		log.Printf("File detected as binary: %s", filePath)
		return map[string]string{"error": fmt.Sprintf("Cannot read binary file: %s", filePath)}, nil
	}
	
	// Read the file
	content, err := os.ReadFile(filePath)
	if err != nil {
		if os.IsPermission(err) {
			return map[string]string{"error": fmt.Sprintf("Permission denied when reading file: %s", filePath)}, nil
		}
		return map[string]string{"error": fmt.Sprintf("Error reading file: %s", err)}, nil
	}
	
	fileContent := string(content)
	log.Printf("Successfully read file: %s (%d chars)", filePath, len(fileContent))
	
	return FileReadResult{
		File:    filePath,
		Content: fileContent,
	}, nil
}

// loadGitignoreMatcher creates a gitignore matcher from .gitignore file
func loadGitignoreMatcher(directory string) gitignore.GitIgnore {
	gitignorePath := filepath.Join(directory, ".gitignore")
	
	// Try to load from file
	matcher, err := gitignore.NewFromFile(gitignorePath)
	if err != nil {
		// If no .gitignore file, create empty matcher
		// For now, we'll return nil and handle it in the caller
		log.Printf("No .gitignore found: %v", err)
		return nil
	} else {
		log.Printf("Loaded gitignore patterns from %s", gitignorePath)
	}
	
	return matcher
}

// shouldIgnore checks if a file should be ignored based on gitignore patterns
// This function works around several issues in the go-gitignore library:
// 1. The library doesn't handle directory patterns correctly (e.g., "node_modules/")
// 2. The library's Match() method can cause nil pointer panics
// 3. The library doesn't work well when not in the repository directory
func shouldIgnore(relPath string, matcher gitignore.GitIgnore) bool {
	if matcher == nil {
		return false
	}
	
	// First try the matcher's Ignore method
	if matcher.Ignore(relPath) {
		return true
	}
	
	// The go-gitignore library has issues with directory patterns.
	// Check if the file is in a directory that should be ignored.
	parts := strings.Split(relPath, string(filepath.Separator))
	for i := 1; i <= len(parts); i++ {
		dirPath := strings.Join(parts[:i], string(filepath.Separator))
		// Check both with and without trailing slash
		if matcher.Ignore(dirPath) || matcher.Ignore(dirPath + "/") {
			return true
		}
	}
	
	return false
}



// isBinary checks if a file is binary by reading the first few bytes
func isBinary(filePath string) bool {
	file, err := os.Open(filePath)
	if err != nil {
		return true // Assume binary if we can't open
	}
	defer file.Close()
	
	// Read first 512 bytes
	buffer := make([]byte, 512)
	n, err := file.Read(buffer)
	if err != nil && err != io.EOF {
		return true
	}
	
	// Check for null bytes (common in binary files)
	for i := 0; i < n; i++ {
		if buffer[i] == 0 {
			return true
		}
	}
	
	// Check if it's mostly printable ASCII
	printable := 0
	for i := 0; i < n; i++ {
		if buffer[i] >= 32 && buffer[i] <= 126 || buffer[i] == '\n' || buffer[i] == '\r' || buffer[i] == '\t' {
			printable++
		}
	}
	
	// If less than 80% printable, consider it binary
	return float64(printable)/float64(n) < 0.8
}

// ExecuteTool executes a tool by name with the given arguments
func ExecuteTool(toolName string, args map[string]interface{}) (string, error) {
	tool, exists := Tools[toolName]
	if !exists {
		return "", fmt.Errorf("unknown tool: %s", toolName)
	}
	
	result, err := tool.Function(args)
	if err != nil {
		return "", err
	}
	
	// Convert result to JSON string
	jsonBytes, err := json.MarshalIndent(result, "", "  ")
	if err != nil {
		return "", fmt.Errorf("error marshaling result: %w", err)
	}
	
	return string(jsonBytes), nil
}

File: noframework/python/common/tools.py
from typing import List, Dict, Any, Union
from pathlib import Path
from .logging import logger
from binaryornot.check import is_binary
from .utils import get_gitignore_spec

# Tool functions
def find_all_matching_files(
    directory: str, 
    pattern: str = "*", 
    respect_gitignore: bool = True, 
    include_hidden: bool = False,
    include_subdirs: bool = True,
    return_paths_as: str = "Path"
    ) -> List[Union[Path, str]]:
    """
    Find files matching a pattern while respecting .gitignore.
    
    Args:
        directory: Directory to search in
        pattern: File pattern to match (glob format)
        respect_gitignore: Whether to respect .gitignore patterns
        include_hidden: Whether to include hidden files and directories
        include_subdirs: Whether to include files in subdirectories
        return_paths_as: Return type for paths - "Path" for Path objects, "str" for strings
        
    Returns:
        List of Path objects or strings for matching files
    """
    logger.info(f"Tool invoked: find_all_matching_files(directory='{directory}', pattern='{pattern}', respect_gitignore={respect_gitignore}, include_hidden={include_hidden}, include_subdirs={include_subdirs})")
    
    try:
        directory_path = Path(directory).resolve()
        logger.debug(f"Resolved directory path: {directory_path}")
        
        if not directory_path.exists():
            logger.warning(f"Directory not found: {directory}")
            return []
        
        # Get gitignore spec if needed
        spec = get_gitignore_spec(str(directory_path)) if respect_gitignore else None
        if spec:
            logger.debug(f"Loaded .gitignore patterns from {directory_path}")
        else:
            logger.debug("No .gitignore patterns loaded (respect_gitignore=False or no .gitignore file)")
        
        result = []
        
        # Choose between recursive and non-recursive search
        if include_subdirs:
            logger.debug(f"Using recursive search (rglob) with pattern: {pattern}")
            paths = directory_path.rglob(pattern)
        else:
            logger.debug(f"Using non-recursive search (glob) with pattern: {pattern}")
            paths = directory_path.glob(pattern)
            
        for path in paths:
            if path.is_file():
                # Skip hidden files if not explicitly included
                # But only skip if they're in hidden directories
                if not include_hidden and path.name.startswith('.'):
                    # Check if any parent directory is hidden (excluding the file itself)
                    rel_path = path.relative_to(directory_path)
                    parent_parts = rel_path.parts[:-1]  # Exclude the filename
                    has_hidden_parent = any(part.startswith('.') for part in parent_parts)
                    
                    # Only skip if it's in a hidden directory, not just a hidden file in root
                    if has_hidden_parent:
                        logger.debug(f"Skipping hidden file in hidden directory: {path}")
                        continue
                    # Hidden files in non-hidden directories (like .gitignore) should be included
                
                # Skip if should be ignored
                if respect_gitignore and spec:
                    # Use pathlib to get relative path and convert to posix format
                    rel_path = path.relative_to(directory_path)
                    rel_path_posix = rel_path.as_posix()
                    if spec.match_file(rel_path_posix):
                        logger.debug(f"Skipping gitignored file: {rel_path_posix}")
                        continue
                result.append(path)
        
        logger.info(f"Found {len(result)} matching files")
        logger.debug(f"Matching files: {[str(p) for p in result[:10]]}{'...' if len(result) > 10 else ''}")
        
        # Return as strings if requested
        if return_paths_as == "str":
            return [str(p) for p in result]
        return result
    except (FileNotFoundError, PermissionError) as e:
        logger.error(f"Error accessing files: {e}")
        return []
    except Exception as e:
        logger.error(f"Unexpected error finding files: {e}")
        return []


# Tool wrapper functions for JSON compatibility
def find_all_matching_files_json(
    directory: str, 
    pattern: str = "*", 
    respect_gitignore: bool = True, 
    include_hidden: bool = False,
    include_subdirs: bool = True
) -> List[str]:
    """
    Wrapper for find_all_matching_files that returns paths as strings for JSON serialization.
    
    This is useful for frameworks that require JSON-serializable outputs (e.g., ADK, LangChain).
    """
    return find_all_matching_files(
        directory=directory,
        pattern=pattern,
        respect_gitignore=respect_gitignore,
        include_hidden=include_hidden,
        include_subdirs=include_subdirs,
        return_paths_as="str"
    )

def read_file(file_path: str) -> Dict[str, Any]:
    """Read the contents of a file."""
    logger.info(f"Tool invoked: read_file(file_path='{file_path}')")
    
    try:
        path = Path(file_path)
        if not path.exists():
            return {"error": f"File not found: {file_path}"}
        
        if is_binary(file_path):
            logger.debug(f"File detected as binary: {file_path}")
            return {"error": f"Cannot read binary file: {file_path}"}
        
        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        file_size = len(content)
        logger.info(f"Successfully read file: {file_path} ({file_size} chars)")
        logger.debug(f"File has {content.count(chr(10))} lines")
        
        return {
            "file": file_path,
            "content": content
        }
    except FileNotFoundError:
        return {"error": f"File not found: {file_path}"}
    except UnicodeDecodeError:
        return {"error": f"Cannot decode file as UTF-8: {file_path}"}
    except PermissionError:
        return {"error": f"Permission denied when reading file: {file_path}"}
    except IOError as e:
        return {"error": f"IO error reading file: {str(e)}"}
    except Exception as e:
        return {"error": f"Unexpected error reading file: {str(e)}"}

# Dictionary mapping tool names to their functions
TOOLS = {
    "find_all_matching_files": find_all_matching_files,
    "read_file": read_file,
}

TOOLS_JSON = {
    "find_all_matching_files": find_all_matching_files_json,
    "read_file": read_file,
}
